---
title: "BCB_Project"
author: "chiteri"
date: "4/4/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
suppressMessages(library(tidyverse))
```

```{r}
#Set working directory
setwd('.')
getwd()
list.files()
#Read data
rna <- read_csv('./GSE121039_AR_scRNAseq_transcript_counts.csv')
#str(rna)
head(rna)
#rename first column

rna <- rename(rna, gene = X1)

#get into long format
rna1 <- pivot_longer(rna, cols = starts_with('Cell'))

#separate name into 4 columns
#rename the columns
#select the columns


rna3 <- separate(rna1, name, into = paste0("y",1:4), sep = " ") %>% 
  rename(c('cell' = y2, 'rep'=y4, 'count' = value)) %>% select(c(gene, cell, rep, count))

##convert cell & rep into numeric & Count unique cells.
n.cells <- unique(as.numeric(rna3$cell))
min(n.cells)
max(n.cells)
#We get 206 cells??
#Find which cells are absent?
which(!(seq(min(n.cells), max(n.cells)) %in% n.cells))
#cells missing
#42,95,166,167,168,169,186,187,189


#list of all cells that failed to pass QC test
#obtained from Table1 of paper
#notice the cells missing from n.cells are here!!
failed <- c(8,12,18,22,25,42,52,54,55,62,63,64,66,67,69,73,75,
            76,77,81,82,83,84,86,88,89,90,91,92,93,95,96,106,113,
            121,123,124,127,129,130,140,141,142,143,144,145,146,147,
            148,149,153,157,163,165,166,167,168,169,170,171,172,173,
            176,182,185,186,187,188,189,191,194,216)

#Filter following cells that failed to pass QC
rna4 <- filter(rna3, !cell %in% failed)

##convert cell & rep into numeric & Count unique cells.
n.cells2 <- unique(as.numeric(rna4$cell)) #now we have 144 cells

#rna4 contains all cells that passed QC and were used in the analysis

##dataset to use for e.g. Fig. 2A, Fig S4). We need 128 cells
#pooling cells using sum
rna5 <- rna4 %>%  group_by(gene,cell,) %>% summarise(n=n(), sum = sum(count))

n.cells3 <- unique(as.numeric(rna5$cell))

rna6 <- filter(rna5, n!=1)
n.cells4 <- unique(as.numeric(rna6$cell)) #We get 128 cells


#PCA
#pivot the data wider for PCA
#First join two columns (cell, rep) into one column cell
#Pivot wider
rna7 <- rna4 %>%  unite("cell",c(cell, rep), sep = "_") %>% pivot_wider(names_from = cell, values_from = count, names_prefix='cell')

#Could not get prcomp to work here so I had to find alternatives
install.packages("FactoMineR")
library(FactoMineR)
library(devtools)
install_github("kassambara/factoextra")
library("factoextra")


#rna8 <- rna7 %>%  rename(" " = gene)

#pca
#remove gene title
rna9 <- rna7 %>% remove_rownames %>% column_to_rownames(var = "gene")

# transpose data frame so cells are row and genes are columns
rna9 <- as.data.frame(t(as.matrix(rna9)))




rna.pca <- PCA(rna9, graph=FALSE)
print(rna.pca)

eigenvalues <- rna.pca$eig
head(eigenvalues[, 1:2])

to_plot <- head(eigenvalues[, 1:2])

#plot eigenvalues vs variance
barplot(to_plot[,2], names.arg=1:nrow(to_plot), 
       main = "Variances",
       xlab = "Principal Components",
       ylab = "Percentage of variances",
       col ="steelblue") +
# Add connected line segments to the plot
lines(x = 1:nrow(to_plot), to_plot[, 2], 
      type="b", pch=19, col = "red")


fviz_screeplot(rna.pca, ncp=10)
```
```{r}

#plot.PCA(rna.pca, axes = c(1,2), choix=c("ind", "var"))

fviz_pca_var(rna.pca, col.var="contrib")
```

```{r}
rna.pca2 <- prcomp(rna7[,2:ncol(rna7)], center=TRUE, scale.=TRUE)

summary(rna.pca2)
```
```{r}
library(devtools)
#install_github("vqv/ggbiplot")
library(ggbiplot)

ggbiplot(rna.pca2)
```
#Figures assignment
Fig 1c
Fig2 a b c d
Fig3 a b d
Fig4 a b c d e
Fig5 c e

##Sean 3 a b d
```{r}

```


##Tanner 2 a c d
```{r}

```

##Yuru 4a,c,5c
```{r}

### number of genes in figure 4
439+570+399+230+781+741



```
## figure 4
```{r}

PCA = prcomp(rna_final, center=TRUE, scale.=TRUE)
summary(PCA)

PCA$x %>% 
  as.data.frame %>%
  ggplot(aes(x = PC1, y = PC2)) + 
  geom_point()


Table_S2 <- read_excel("aav6428_Nelms_Table-S2.xlsx")
#str(Table_S2)
colnames(Table_S2) <- c("gene", "Cell_cycle_phase",  "Meiotic_expression_level", "Gene_discription")
Table_S2$Cell_cycle_phase <- as.factor(Table_S2$Cell_cycle_phase)
dim(Table_S2)
MEG <- Table_S2$gene ## Expresses genes in meiotic cycle 
length(MEG)

Table_S3 <- read_excel("aav6428_Nelms_Table-S3.xlsx")
str(Table_S3)
colnames(Table_S3) <- c("Gene", "Gene_score",  "Cluster", "Transform_value", "Meiotic_gene", "TF", "Gene_discription")
Table_S3$Cluster <- as.factor(Table_S3$Cluster)
DEG <- Table_S3$Gene # differentially expressed genes
length(DEG)


rna_norm %>% rownames_to_column(var = "Gene")%>%left_join(Table_S3)  %>%  
  ggplot()+
  geom_bar(aes(x=Cluster),na.rm = T)



Table_S3 %>% filter(Gene %in% MEG) %>% 
  ggplot()+
  geom_bar(aes(x=Cluster))

Table_S3 %>% filter(!Gene %in% MEG) %>% 
  ggplot()+
  geom_bar(aes(x=Cluster))


```


## Figure 5a

```{r}



rna_ftf

Test <- rna_norm %>% rownames_to_column(var = "gene")%>% right_join(Table_S2[,1]) #%>% drop_na()
rownames(Test) <- Test$gene
Test <- Test[,-1]
dim(Test)

rna_norm
dim(rna_norm)
rna_norm_3046 <- rna_norm %>% rownames_to_column(var = "Gene") %>% filter(Gene %in% DEG) %>% select(!Gene)
dim(rna_norm_3046)

typeof(MEG)

rna_norm_456 <- rna_norm1 %>% rownames_to_column(var = "Gene") %>% filter(Gene %in% MEG) #%>% select(!Gene)
dim(rna_norm_456)

rownames(rna_norm)
PCA1 = prcomp(rna_norm_3046, center=TRUE, scale.=TRUE)
summary(PCA1)

PCA2 = prcomp(rna_norm_456[,-1], center=TRUE, scale.=TRUE)
summary(PCA2)



Table_S2.1 <- Table_S2 %>% filter (gene %in% rna_norm_456$Gene)
dim(Table_S2.1)
dim(PCA2$x)

rna_norm_456
PCA2$x %>% 
  as.data.frame %>%mutate(phase=Table_S2.1$Cell_cycle_phase) %>% 
  ggplot(aes(x = PC1, y = PC2)) + 
  geom_point(aes(col=phase))

#rna4 %>%left_join(Table_S1.1) %>% left_join(Table_S2[,1:2]) %>% drop_na()

```




## Figure 5C
```{r}
library(readxl)
Table_S1 <- read_excel("aav6428_Nelms_Table-S1.xlsx",sheet = "Table S1")
dim(Table_S1)
colnames(Table_S1)
Table_S1.1 <- Table_S1[,c(1, 3, 13, 14)]
colnames(Table_S1.1) <- c("cell", "Anther_length", "Expression_stage", "Cell_cycle_phase")
Table_S1.1$cell <- as.factor(Table_S1.1$cell)
Table_S1.1$Cell_cycle_phase <- as.factor(Table_S1.1$Cell_cycle_phase)

cbp1 <- c("#999999", "#56B4E9", "#C3D7A4","#009E73","#293352", "#0072B2")
Table_S1.1 %>% filter(!Cell_cycle_phase == "N.D.") %>% droplevels() %>% 
  mutate(anther_length = cut(Anther_length, breaks= c(0.2, 0.4, 0.6, 0.85, 1.1, 1.25, 1.4, 1.6))) %>%
  ggplot()+
  geom_bar(aes(x=anther_length, fill=Cell_cycle_phase), position = "fill", width=0.5)+
  scale_fill_manual(values = cbp1)

```



##Chiteri 1 c 2 b 5e
```{r}

```

##Juan 4b d e
```{r}

qc <- readxl::read_excel('./aav6428_Nelms_Table-S1.xlsx',sheet = "Table S1" )
names(qc)[12] <- 'QC' 
qc <- qc[qc$QC == "PASS", "Cell Number"]

deg <- readxl::read_excel('./aav6428_Nelms_Table-S2.xlsx')[,1:2]
names(deg) <- c('gene', 'phase')
deg <- deg %>% filter(phase != "G1") %>% select(gene)

rna4_test <- filter(rna3, cell %in% qc$`Cell Number`)
rna6_test <- filter(rna5, n > 1)

rna7_test <- rna4_test %>%  
  unite("cell",c(cell, rep), sep = "_") %>% 
  pivot_wider(names_from = cell, values_from = count, names_prefix = 'cell')

rna9_t <- as.data.frame(t(as.matrix(rna9)))

cells <- as.data.frame(row.names(rna9_t))
names(cells) <- 'key'
cells %>% 
  separate(key, c('cell', 'rep'), remove = F) %>% 
  group_by(cell) %>% filter(n() == 2) %>% 
  ungroup() %>% select(key) -> cells_after_qc

rna9_t %>% rownames_to_column() %>% 
  filter(rowname %in% cells_after_qc$key) %>% 
  column_to_rownames() -> rna9_f

rna9_ft <- as.data.frame(t(as.matrix(rna9_t)))
rna9_ft %>% rownames_to_column("gene") -> rna9_ft

rna9_ft %>%   
  arrange(gene) %>% rowwise() %>%  
  group_by(gene) %>% dplyr::summarise(m = sum(c_across(2:length(rna9_ft)-1))) %>%  
  filter(m > 100) %>% select(gene) -> exp_genes
rna_ftf <- as.tibble(rna9_ft[rna9_ft$gene %in% exp_genes$gene,])
dim(rna_ftf)

rna_ftfc <- rna_ftf %>% filter(!gene %in% deg$gene)

rna_norm <- as.tibble(rna_ftfc$gene) # tibble ngenes x 1 for starting fillling with normalized data
for (i in 2:length(rna_ftfc)) {
  a <- (rna_ftfc[,i]/sum(rna_ftfc[,i]))*1000000
  rna_norm <- cbind(rna_norm, a)
  a <- NULL
}
rna_norm <- column_to_rownames(rna_norm, 'value') # for transposing
# Step 2: log transformation after add 11
rna_norm <- log10(rna_norm[,1:length(rna_norm)] + 11 )
rna_norm[1:5, 1:5]
rna_final <- as.data.frame(t(as.matrix(rna_norm)))
rna_final[1:5, 1:5]

write_delim(rna_final, "rna_final_4PCA.txt", delim='\t')

dim(rna_final)

PCA = prcomp(rna_final)
summary(PCA)
PCA$x %>% 
  as.data.frame %>%
  ggplot(aes(x = PC1, y = PC2)) + 
  geom_point()



TableS3 %>% arrange(desc("Gene score (% variance explained by pseudotime)")) %>% 
  select(1:2) %>% 
  top_n(2000)



```


