---
title: "Figure 3B SM"
author: "Sean McLaughlin"
date: "4/28/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
#Code for PCA & Pseudotime
```{r}
library("tidyverse")
library("factoextra")
library("FactoMineR")
library("readxl")
library("data.table")
library("princurve")

suppressPackageStartupMessages(library("tidyverse"))
suppressPackageStartupMessages(library("factoextra"))
suppressPackageStartupMessages(library("FactoMineR"))
suppressPackageStartupMessages(library("ggfortify"))



```

```{r}
#Set working directory
setwd('.')
getwd()
#list.files()
```

#Reading data

```{r}
# Counts
rna <- read_csv('./GSE121039_AR_scRNAseq_transcript_counts.csv')
rna <- rename(rna, gene = X1) #rename first column

# Quality control: PASS/FAIL


qc <- readxl::read_excel('./aav6428_Nelms_Table-S1.xlsx', sheet = 'Table S1')



qc <- readxl::read_excel('./aav6428_Nelms_Table-S1.xlsx', sheet = "Table S1")

qc <- readxl::read_excel('./aav6428_Nelms_Table-S1.xlsx', sheet = 'Table S1')


names(qc)[12] <- 'QC' 
qc <- qc[qc$QC == "PASS", "Cell Number"]

# Deferentially expressed genes during specific stages
deg <- readxl::read_excel('./aav6428_Nelms_Table-S2.xlsx')[,1:2]
names(deg) <- c('gene', 'phase')
deg <- deg %>% filter(phase != "G1") %>% select(gene)

head(rna)
head(qc)
```

```{r}
#get into long format
rna1 <- pivot_longer(rna, cols = starts_with('Cell'))
```
#rename the columns
#select the columns


```{r}
#separate 'name' into 4 columns, 

rna3 <- separate(rna1, name, into = paste0("y",1:4), sep = " ") %>% 
  select(!c(y1, y3)) %>% rename(c('cell' = y2, 'rep'= y4, 'count' = value))

```


```{r}
#Filter cells based on QC
rna4 <- filter(rna3, cell %in% qc$`Cell Number`)
dim(rna4)
head(rna4)

paste0("Total number of cells after Quality Control: ", length(unique(as.numeric(rna4$cell))))
```

``` {r}

##dataset to use for e.g. Fig. 2A, Fig S4). We need 128 cells
#pooling cells using sum
rna5 <- rna4 %>%  group_by(gene,cell,) %>% summarise(n=n(), sum = sum(count))
paste("There are ", length(unique(as.numeric(rna5$cell))), " unique cells in rna5")
```

``` {r}
rna6 <- filter(rna5, n > 1)

paste0("There are ", length(unique(as.numeric(rna6$cell))), " cells with a technical replicate in rna6.")
```


``` {r}
#Join two columns cell and rep and go wider

rna7 <- rna4 %>%  
  unite("cell",c(cell, rep), sep = "_") %>% 
  pivot_wider(names_from = cell, values_from = count, names_prefix = 'cell')

dim(rna7)
```


``` {r}
#remove gene title
rna9 <- rna7 %>% remove_rownames %>% column_to_rownames(var = "gene")

dim(rna9)

rna7[1:5, 1:5]
rna9[1:5, 1:4]

```

```{r}
# transpose data frame so cells are rows and genes are columns
rna9_t <- as.data.frame(t(as.matrix(rna9)))
```

```{r}
# next two chunks for getting rid of 'unreplicated' cells 
cells <- as.data.frame(row.names(rna9_t))
names(cells) <- 'key'

cells %>% 
  separate(key, c('cell', 'rep'), remove = F) %>% 
  group_by(cell) %>% filter(n() == 2) %>% 
  ungroup() %>% select(key) -> cells_after_qc
```


```{r}
# filtering out rna9 after quality control
rna9_t %>% rownames_to_column() %>% 
  filter(rowname %in% cells_after_qc$key) %>% 
  column_to_rownames() -> rna9_f
```


```{r}
# Re arranging rna9_aqc moving cells as columns 
rna9_ft <- as.data.frame(t(as.matrix(rna9_f)))
rna9_ft %>% rownames_to_column("gene") -> rna9_ft 
```

```{r}
# Filtering out genes with less than 100 counts across cells 
rna9_ft %>%   
  arrange(gene) %>% rowwise() %>%  
  group_by(gene) %>% dplyr::summarise(m = sum(c_across(2:length(rna9_ft)-1))) %>%  
  filter(m > 100) %>% select(gene) -> exp_genes

rna_ftf <- as_tibble(rna9_ft[rna9_ft$gene %in% exp_genes$gene,])
dim(rna_ftf)
```

#Filtering out genes differentially expressed in specific stages
```{r}
rna_ftfc <- rna_ftf %>% filter(!gene %in% deg$gene)
```

Normalization
```{r}

# Step 1: divided each individual count by the total counts per cell, and multiply by 1000000 

rna_norm <- as_tibble(rna_ftfc$gene) # tibble ngenes x 1 for starting fillling with normalized data

for (i in 2:length(rna_ftfc)) {
  a <- (rna_ftfc[,i]/sum(rna_ftfc[,i]))*1000000
  rna_norm <- cbind(rna_norm, a)
  a <- NULL
}

rna_norm <- column_to_rownames(rna_norm, 'value') # for transposing

# Step 2: log transformation after add 11

rna_norm <- log10(rna_norm[,1:length(rna_norm)] + 11 )

rna_norm[1:5, 1:5]

rna_final <- as.data.frame(t(as.matrix(rna_norm)))
rna_final[1:5, 1:5]

```


```{r}
PCA = prcomp(rna_final, center=TRUE, scale.=FALSE)
summary(PCA)

PCA$x %>% 
  as.data.frame %>%
  ggplot(aes(x = PC1, y = PC2)) + 
  geom_point()

```

```{r}

#Pseudotime

#Take Juan's Normalized genes and the filter with 2000 genes with highest variance and use for pseudotime. From fig. S3

 
  ##PCA for pairwise comparison of replicates


 Anther_size = read.table("Updated_S1.txt", header= T)

####Transform data to merge files to use only 2000 genes
rna_final_Tanner= as.data.frame(t(rna_final)) 
s3= read.csv("aav6428_Nelms_Table-S3.csv", header=T)

#Add a column label to rownames
setDT(rna_final_Tanner, keep.rownames = "gene")[]

#Merge Files
Variance2000transcripts=inner_join(rna_final_Tanner,s3 )### MAke a combined table of all fig s3, order by variance and then slice 2000

y= Variance2000transcripts %>% arrange(desc("Gene score (% variance explained by pseudotime)")) %>% 
  select(1:258) %>% 
  top_n(2000)

 
#Remove gene id column
y$gene=NULL

#Trying the PCA with prcomp

PCA2= prcomp(y, center=TRUE, scale.=FALSE)
summary(PCA)
PCA2$x %>% 
  as.data.frame %>%
  ggplot(aes(x = PC1, y = PC2)) + 
  geom_point()

plot(PCA2)
names(PCA2)


#PCA 10 as a matrix while selecting the X values from the PCA
PCAX= as.matrix(PCA2$x [,1:10])
PCAX

I=principal_curve(PCAX)
print(I)
#Principal Curve
plot(I, xmax=10, ymax=-20)
plot

plot(I$lambda)
plot(PCA2$x) # any of the PCA names can be called here (where x is)


#PC1vsPC2 graph

J=PCA2$x %>% 
  as.data.frame %>%
  ggplot(aes(x=PC1,y=PC2)) + geom_point(size=1) +
  theme_bw(base_size=32)
J

# "X" is the principal components of interest *Source:https://cmdlinetips.com/2019/04/introduction-to-pca-with-r-using-prcomp/

#https://jokergoo.github.io/ComplexHeatmap-reference/book/



######
```


#Figure 3A
```{r}
library(readxl)
aav6428_Nelms_Table_S1 <- read_excel("aav6428_Nelms_Table-S1.xlsx", sheet = "Table S1")
View(aav6428_Nelms_Table_S1)

library(readr)
GSE121039_AR_scRNAseq_transcript_counts <- read_csv("GSE121039_AR_scRNAseq_transcript_counts.csv")
View(GSE121039_AR_scRNAseq_transcript_counts)
library(dplyr)

```

#Subset Marker Genes from RNA Expression data as Datasets
```{r}
#Subsetting Marker Genes by ZM Name - RPL-38e cannot be found
ago18a <- GSE121039_AR_scRNAseq_transcript_counts[ GSE121039_AR_scRNAseq_transcript_counts$"X1" %in% c("Zm00001d006351"), ]

RMF <- GSE121039_AR_scRNAseq_transcript_counts[ GSE121039_AR_scRNAseq_transcript_counts$"X1" %in% c("Zm00001d008945"), ]

C3H3 <- GSE121039_AR_scRNAseq_transcript_counts[ GSE121039_AR_scRNAseq_transcript_counts$"X1" %in% c("Zm00001d039101"), ]

trps8 <- GSE121039_AR_scRNAseq_transcript_counts[ GSE121039_AR_scRNAseq_transcript_counts$"X1" %in% c("Zm00001d050069"), ]
```

#Attempting to Graph Marker Gene datasets with Pseudotime
```{r}
M_Mark_Psu1 <- merge()#??
#How to merge data and plot on same graph?
ggplot() +
  geom_point(data = c(ago18a,I$s), aes(I$s,1), color = "blue")

```
#Figure 3B
```{r}
#Remove and arrange anther data
ANTHER1=read.csv("with_anther_size", header=T)


ANTHER2=ANTHER1[order(ANTHER1$anther_size_mm),]
ANTHER2=as.tibble(ANTHER2)

#Remove cells that don't pass QC Result=144 cells
ANTHER3 <- filter(ANTHER2, cell %in% qc$`Cell Number`)
ANTHER4=t(ANTHER3)

#Remove rows X, cell, rep,anther_size_mm
ANTHER5=ANTHER4[-c(1,2,3,4),]
ANTHER6=as.data.frame(ANTHER5)

#Subset the Marker Genes
ANTH_MARKER <- ANTHER6[ANTHER6$"gene" %in% c("Zm00001d006351","Zm00001d008945","Zm00001d039101","Zm00001d050069"), ]

#Add row names as a column for sorting for each gene

setDT(ANTH_MARKER, keep.rownames = "gene")[]

#Anther length with Marker Genes
Anther_length_w_M=inner_join(ANTH_MARKER,s3)

#Remove gene names 
Anther_length_w_M$gene=NULL

#Heatmap the three genes
vst_cor_M<- as.matrix(cor(Anther_length_w_M, method="spearman"))
pheatmap(Anther_length_w_M,border_color=NA,labels_row = FALSE,labels_col = F)

```


