---
title: "BCB_Project"
author: "Chiteri and Tanner and Juan"
date: "4/4/2021"
output: html_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
#Dependencies

library("tidyverse")
library("factoextra")
library("FactoMineR")
library("readxl")

suppressPackageStartupMessages(library("tidyverse"))
suppressPackageStartupMessages(library("factoextra"))
suppressPackageStartupMessages(library("FactoMineR"))
suppressPackageStartupMessages(library("ggfortify"))



```

```{r}
#Set working directory
setwd('.')
getwd()
#list.files()
```

#Reading data

```{r}
# Counts
rna <- read_csv('./GSE121039_AR_scRNAseq_transcript_counts.csv')
rna <- rename(rna, gene = X1) #rename first column

# Quality control: PASS/FAIL


qc <- readxl::read_excel('./aav6428_Nelms_Table-S1.xlsx', sheet = 'Table S1')



qc <- readxl::read_excel('./aav6428_Nelms_Table-S1.xlsx', sheet = "Table S1")

qc <- readxl::read_excel('./aav6428_Nelms_Table-S1.xlsx', sheet = 'Table S1')


names(qc)[12] <- 'QC' 
qc <- qc[qc$QC == "PASS", "Cell Number"]

# Deferentially expressed genes during specific stages
deg <- readxl::read_excel('./aav6428_Nelms_Table-S2.xlsx')[,1:2]
names(deg) <- c('gene', 'phase')
deg <- deg %>% filter(phase != "G1") %>% select(gene)

head(rna)
head(qc)
```

```{r}
#get into long format
rna1 <- pivot_longer(rna, cols = starts_with('Cell'))
```
#rename the columns
#select the columns


```{r}
#separate 'name' into 4 columns, 

rna3 <- separate(rna1, name, into = paste0("y",1:4), sep = " ") %>% 
  select(!c(y1, y3)) %>% rename(c('cell' = y2, 'rep'= y4, 'count' = value))

```


```{r}
#Filter cells based on QC
rna4 <- filter(rna3, cell %in% qc$`Cell Number`)
dim(rna4)
head(rna4)

paste0("Total number of cells after Quality Control: ", length(unique(as.numeric(rna4$cell))))
```

``` {r}

##dataset to use for e.g. Fig. 2A, Fig S4). We need 128 cells
#pooling cells using sum
rna5 <- rna4 %>%  group_by(gene,cell,) %>% summarise(n=n(), sum = sum(count))
paste("There are ", length(unique(as.numeric(rna5$cell))), " unique cells in rna5")
```

``` {r}
rna6 <- filter(rna5, n > 1)

paste0("There are ", length(unique(as.numeric(rna6$cell))), " cells with a technical replicate in rna6.")
```


``` {r}
#Join two columns cell and rep and go wider

rna7 <- rna4 %>%  
  unite("cell",c(cell, rep), sep = "_") %>% 
  pivot_wider(names_from = cell, values_from = count, names_prefix = 'cell')

dim(rna7)
```


``` {r}
#remove gene title
rna9 <- rna7 %>% remove_rownames %>% column_to_rownames(var = "gene")

dim(rna9)

rna7[1:5, 1:5]
rna9[1:5, 1:4]

```

```{r}
# transpose data frame so cells are rows and genes are columns
rna9_t <- as.data.frame(t(as.matrix(rna9)))
```

```{r}
# next two chunks for getting rid of 'unreplicated' cells 
cells <- as.data.frame(row.names(rna9_t))
names(cells) <- 'key'

cells %>% 
  separate(key, c('cell', 'rep'), remove = F) %>% 
  group_by(cell) %>% filter(n() == 2) %>% 
  ungroup() %>% select(key) -> cells_after_qc
```


```{r}
# filtering out rna9 after quality control
rna9_t %>% rownames_to_column() %>% 
  filter(rowname %in% cells_after_qc$key) %>% 
  column_to_rownames() -> rna9_f
```


```{r}
# Re arranging rna9_aqc moving cells as columns 
rna9_ft <- as.data.frame(t(as.matrix(rna9_f)))
rna9_ft %>% rownames_to_column("gene") -> rna9_ft 
```

```{r}
# Filtering out genes with less than 100 counts across cells 
rna9_ft %>%   
  arrange(gene) %>% rowwise() %>%  
  group_by(gene) %>% dplyr::summarise(m = sum(c_across(2:length(rna9_ft)-1))) %>%  
  filter(m > 100) %>% select(gene) -> exp_genes

rna_ftf <- as_tibble(rna9_ft[rna9_ft$gene %in% exp_genes$gene,])
dim(rna_ftf)
```

#Filtering out genes differentially expressed in specific stages
```{r}
rna_ftfc <- rna_ftf %>% filter(!gene %in% deg$gene)
```

Normalization
```{r}

# Step 1: divided each individual count by the total counts per cell, and multiply by 1000000 

rna_norm <- as_tibble(rna_ftfc$gene) # tibble ngenes x 1 for starting fillling with normalized data

for (i in 2:length(rna_ftfc)) {
  a <- (rna_ftfc[,i]/sum(rna_ftfc[,i]))*1000000
  rna_norm <- cbind(rna_norm, a)
  a <- NULL
}

rna_norm <- column_to_rownames(rna_norm, 'value') # for transposing

# Step 2: log transformation after add 11

rna_norm <- log10(rna_norm[,1:length(rna_norm)] + 11 )

rna_norm[1:5, 1:5]

rna_final <- as.data.frame(t(as.matrix(rna_norm)))
rna_final[1:5, 1:5]

```


```{r}
PCA = prcomp(rna_final, center=TRUE, scale.=FALSE)
summary(PCA)

PCA$x %>% 
  as.data.frame %>%
  ggplot(aes(x = PC1, y = PC2)) + 
  geom_point()

ggsave('pca2.jpeg')

```

#Figures assignment
Fig 1c
Fig2 a b c d
Fig3 a b d
Fig4 a b c d e
Fig5 c e

##Sean 3 a b d
```{r}

```


##Tanner 2 a c d


```{r}
library(ggplot2)
library("princurve")
library(dplyr)
library(data.table)
library(pheatmap)

```


#Pseudotime

#Take Juan's Normalized genes and the filter with 2000 genes with highest variance and use for pseudotime. From fig. S3



```{r}
library(ggplot2)
library("princurve")
library(dplyr)
library(data.table)
library(pheatmap)

#Pseudotime

#Take Juan's Normalized genes and the filter with 2000 genes with highest variance and use for pseudotime. From fig. S3

 
 library("princurve")
 install.packages('PRcurve')
 suppressWarnings(suppressPackageStartupMessages(library(ggplot2)))
 suppressWarnings(suppressPackageStartupMessages(library(cowplot)))

 if (!requireNamespace("BiocManager", quietly = TRUE))
     install.packages("BiocManager")
 library(slingshot)

 library(monocle)
 library("princurve")
 library(dplyr)

 ##PCA for pairwise comparison of replicates

 library(readr)
 library(tidyverse)

 Anther_size = read.table("Updated_S1.txt", header= T)

####Transform data to merge files to use only 2000 genes
rna_final_Tanner= as.data.frame(t(rna_final)) 
s3= read.csv("aav6428_Nelms_Table-S3.csv", header=T)

#Add a column label to rownames
setDT(rna_final_Tanner, keep.rownames = "gene")[]

#Merge Files
Variance2000transcripts=inner_join(rna_final_Tanner,s3 )### MAke a combined table of all fig s3, order by variance and then slice 2000

y= Variance2000transcripts %>% arrange(desc("Gene score (% variance explained by pseudotime)")) %>% 
  select(1:258) %>% 
  top_n(2000)

 
#Remove gene id column
y$gene=NULL

#Trying the PCA with prcomp

PCA2= prcomp(y, center=TRUE, scale.=FALSE)
summary(PCA)
PCA2$x %>% 
  as.data.frame %>%
  ggplot(aes(x = PC1, y = PC2)) + 
  geom_point()

plot(PCA2)
names(PCA2)


#PCA 10 as a matrix while selecting the X values from the PCA
PCAX= as.matrix(PCA2$x [,1:10])
PCAX

I=principal_curve(PCAX)
print(I)
#Principal Curve
plot(I)
plot(I$lambda)
plot(PCA2$x) # any of the PCA names can be called here (where x is)

#PC1vsPC2 graph

J=PCA2$x %>% 
  as.data.frame %>%
  ggplot(aes(x=PC1,y=PC2)) + geom_point(size=1) +
  theme_bw(base_size=32)
J


 # "X" is the principal components of interest *Source:https://cmdlinetips.com/2019/04/introduction-to-pca-with-r-using-prcomp/



### Heat map...
#https://jokergoo.github.io/ComplexHeatmap-reference/book/
y
y$gene<-NULL
vst_cor <- as.matrix(cor(y, method="spearman")) ## 2000 genes 144 Cells



pheatmap(vst_cor,border_color=NA,labels_row = FALSE, labels_col = F)



######

## Heatmap 
#Sort by Anther size
ANTHER1=read.csv("with_anther_size", header=T)


ANTHER2=ANTHER1[order(ANTHER1$anther_size_mm),]
ANTHER3=t(ANTHER2)

ANTHER4=ANTHER3[-c(1,2,3,4),]
ANTHER5=as.data.frame(ANTHER4)

#Add row names as a column for sorting
library(data.table)

setDT(ANTHER5, keep.rownames = "gene")[]

#Anther length with 3040 genes
Anther_length_w_genes=inner_join(ANTHER5,s3) # 3040 genes and 273 cells, too many

#Remove gene names 
Anther_length_w_genes$gene=NULL


vst_cor2 <- as.matrix(cor(Anther_length_w_genes, method="spearman"))
pheatmap(vst_cor2,border_color=NA,labels_row = FALSE, labels_col = F)

#
 #Identify 2000 genes with highest variance and use for pseudotime. From fig. S3
 variance2000 = read.table("2000genE_s3.txt", header= T)

 Variance2000transcripts=right_join(rna7,variance2000, "gene")

 #Trying the PCA with prcomp
 #remove gene title
 rna10 <- Variance2000transcripts %>% remove_rownames %>% column_to_rownames(var = "gene")
 PCA= prcomp(rna10[, c(2:272)], center=TRUE, scale.=TRUE)
 summary(PCA)


```
 










```{r include=FALSE}
##Everything below doesn't help out but please DO NOT DELETE

##########################################################################################################################


#Adding Anther size... it may not matter

###################################################
# make the columns the same value type
#Anther_sort$cell <- as.numeric(Anther_sort$cell)
#Anther_sort$rep <- as.numeric(Anther_sort$rep)
#Anther_sort$count <- as.numeric(Anther_sort$count)
 
library(readr)

Anther_size = read.table("Updated_S1.txt", header= T)


rna4$cell <- as.numeric(as.character(rna4$cell))

Anther_sort=right_join(rna4, Anther_size, "cell")

Anther_sorted=Anther_sort[with(Anther_sort,order(Anther_size)), ]


#REMOVE UNWANTED COLUMNS
rnaT <- select(Anther_sorted, -c(5:6)) 

rna7T <- rnaT %>%  unite("cell",c(cell, rep), sep = "_") %>% pivot_wider(names_from = cell, values_from = count, names_prefix='cell')

dim(rna7T)




##################################


 ####
 library(stringr)
 library(dplyr)

 ls()
 rm(list=setdiff(ls(), "rna4"))

 check_reps <- rna4 %>% group_by(gene, cell) %>% summarise(n = n())
 head(check_reps)
 unique(check_reps$n)
 t
 str(rna4)

 check_reps <- check_reps %>% filter(n == 1)

####
#MAtheus help library(stringr)
library(dplyr)

install.packages('LearnBayes')



HHHHH <- rna4 %>% group_by(cell, rep) %>% filter( rep =="2") %>% droplevels()

y= as.data.frame(HHHHH)
x= as.data.frame(rna4)

JoinHHHH= right_join(y, x, "cell")
JJJJJ <- rna4 %>% select(rep) %>% filter( rep =="2")%>% droplevels()

a= c(HHHHH)
b=c(JJJJJ)

df= data.frame(a,b)

c
#Using Monocle


#########################
#MAtheus help library(stringr)
library(dplyr)

ls()


 ####
 library(stringr)
 library(dplyr)

 ls()
 rm(list=setdiff(ls(), "rna4"))


 check_reps <- rna4 %>% group_by(gene, cell) %>% summarise(n = n())
 head(check_reps)
 unique(check_reps$n)
 t
 str(rna4)


 to_remove <- paste0(check_reps$gene, check_reps$cell)
 to_remove[1:5]

 rna4$index <- paste0(rna4$gene, rna4$cell)

 pca_data <- rna4[-match(to_remove,rna4$index),]

 dim(pca_data)
 dim(rna4)
 head(pca_data)

 check_reps <- check_reps %>% filter(n == 1)


 to_remove <- paste0(check_reps$gene, check_reps$cell)
 to_remove[1:5]

 rna4$index <- paste0(rna4$gene, rna4$cell)

 pca_data <- rna4[-match(to_remove,rna4$index),]


 dim(pca_data)
 dim(rna4)
 head(pca_data)


 check_reps <- pca_data %>% group_by(gene, cell) %>% summarise(n  n())
 check_reps[which(check_reps$n != 2),]


 check_reps <- pca_data %>% group_by(gene, cell) %>% summarise(n  n())
 check_reps[which(check_reps$n != 2),]

 pca_data_final <- data.frame(R1 = pca_data$count[which(pca_data$rep == 1)],
                              R2 = pca_data$count[which(pca_data$rep == 2)])

 sum(pca_data$rep == 1)
 sum(pca_data$rep == 2)

 #Using Monocle

  pca_data_final <- data.frame(R1 = pca_data$count[which(pca_data$rep == 1)],
                              R2 = pca_data$count[which(pca_data$rep == 2)])

 sum(pca_data$rep == 1)
 sum(pca_data$rep == 2)

sum(pca_data$rep == 1)
sum(pca_data$rep == 2)
########################################################################################################
#Please do not delete ^
```



##Yuru 4a,c,5c
```{r}

```

##Chiteri 1 c 2 b 5e
```{r}
# Quality control: PASS/FAIL
table1 <- readxl::read_excel('./aav6428_Nelms_Table-S1.xlsx', sheet = 'Table S1')

#rename the columns..replace spaces with underscore and lower case
table1.1 <- table1 %>%  
  select_all(~gsub("\\s", "_", .)) %>% 
  select_all(tolower)

#rename anthersize column name
colnames(table1.1)[3] <- 'anther_size_mm'
colnames(table1.1)[1] <- 'cell'

#filter only cells that passed

table1.2 <- table1.1 %>%  filter(quality_control=="PASS") %>%  select(cell, anther_size_mm)

#Join two columns cell and rep and go wider

rna4_wide <- rna4 %>% pivot_wider(names_from = gene, values_from = count)

rna4_anther <- merge(rna4_wide, table1.2, by='cell') %>% select(cell, rep, anther_size_mm, everything()) %>% arrange(as.numeric(cell))

rna4_replicated <- rna4_anther %>% group_by(cell) %>% filter(n()==2) %>%  ungroup() 


#make anther_size_mm for grouping purposes
rna4_replicated$anther_size_mm <- as.character(rna4_replicated$anther_size_mm)

#filter whose sum across samples is > 100
rna4_filtered <- rna4_replicated[4:ncol(rna4_replicated)][, apply(rna4_replicated[4:ncol(rna4_replicated)],2,sum) > 100]


#we get 256 obs, 12912 genes, 10 shy of 12902 from the paper

#introduce the cell, rep, anther_size
rna4_x <- cbind(rna4_replicated[1:3], rna4_filtered )

#filter the 375 
# Deferentially expressed genes during specific stages
rna4_x2 <- rna4_x[, !names(rna4_x) %in% deg$gene] 
#we get 12544 against 12527 in the paper. We have 17 more??


#make anther_size_mm for grouping purposes
rna4_x2$anther_size_mm <- as.numeric(rna4_x2$anther_size_mm)

##normalize (uses columns)
#Uses row sums
rna4_x3 <- apply(rna4_x2[4:ncol(rna4_x2)], 1,function(i) log10(((i/sum(i))*1000000)+11 ))

rna4_log3 <- as.data.frame(t(rna4_x3))

#combine
rna4_log3 <- cbind(rna4_x2[,1:3], rna4_log3)

PCA_256 = prcomp(rna4_log3[, -c(1:3)], center=TRUE, scale.=FALSE)
summary(PCA_256)

autoplot(PCA_256, data=rna4_x2, colour= 'anther_size_mm') + geom_line(aes(group = cell))+ scale_color_gradient2(low = "gray",high = "black", space = "Lab" )+theme_bw() 

ggsave('pca2.png')



```
```{r}
##kmeans with cycle
#kmeans plot
set.seed(20)
#autoplot(kmeans(rna4_log3[, -c(1:3)], 6), data=rna4_x2 )

#Producing error currently
ggsave('cell_cycle_cluster.png')


```
```{r}
##kmeans with cycle
#kmeans plot
set.seed(20)
#autoplot(kmeans(rna4_log3[, -c(1:3)], 6), data=rna4_x2 )

#Producing error currently
ggsave('cell_cycle_cluster.png')

```



##Juan 4b d e
```{r}

```

