---
title: "BCB_Project"
author: "Chiteri and Tanner and Juan"
date: "4/4/2021"
output: html_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
#Dependencies

library("tidyverse")



```


#Reading data

```{r}
# Counts
rna <- read_csv('./Data/GSE121039_AR_scRNAseq_transcript_counts.csv')
rna <- rename(rna, gene = X1) #rename first column

# Quality control: PASS/FAIL


qc <- readxl::read_excel('./Data/aav6428_Nelms_Table-S1.xlsx', sheet = 'Table S1')



qc <- readxl::read_excel('./Data/aav6428_Nelms_Table-S1.xlsx', sheet = "Table S1")

qc <- readxl::read_excel('./Data/aav6428_Nelms_Table-S1.xlsx', sheet = 'Table S1')


names(qc)[12] <- 'QC' 
qc <- qc[qc$QC == "PASS", "Cell Number"]

# Deferentially expressed genes during specific stages
deg <- readxl::read_excel('./Data/aav6428_Nelms_Table-S2.xlsx')[,1:2]
names(deg) <- c('gene', 'phase')
deg <- deg %>% filter(phase != "G1") %>% select(gene)

head(rna)
head(qc)
```

```{r}
#get into long format
rna1 <- pivot_longer(rna, cols = starts_with('Cell'))
```
#rename the columns
#select the columns


```{r}
#separate 'name' into 4 columns, 

rna3 <- separate(rna1, name, into = paste0("y",1:4), sep = " ") %>% 
  select(!c(y1, y3)) %>% rename(c('cell' = y2, 'rep'= y4, 'count' = value))

```


```{r}
#Filter cells based on QC
rna4 <- filter(rna3, cell %in% qc$`Cell Number`)
names(rna4)[1] <- 'gene'
head(rna4)

paste0("Total number of cells after Quality Control: ", length(unique(as.numeric(rna4$cell))))
```

``` {r}

##dataset to use for e.g. Fig. 2A, Fig S4). We need 128 cells
#pooling cells using sum
rna5 <- rna4 %>%  group_by(gene,cell,) %>% summarise(n=n(), sum = sum(count))
paste("There are ", length(unique(as.numeric(rna5$cell))), " unique cells in rna5")

```

``` {r}
rna6 <- filter(rna5, n > 1)

paste0("There are ", length(unique(as.numeric(rna6$cell))), " cells with a technical replicate in rna6.")
```


``` {r}
#Join two columns cell and rep and go wider

rna7 <- rna4 %>%  
  unite("cell",c(cell, rep), sep = "_") %>% 
  pivot_wider(names_from = cell, values_from = count, names_prefix = 'cell')

dim(rna7)
```


``` {r}
#remove gene title
rna9 <- rna7 %>% remove_rownames %>% column_to_rownames(var = "gene")

dim(rna9)

rna7[1:5, 1:5]
rna9[1:5, 1:4]

```

```{r}
# transpose data frame so cells are rows and genes are columns
rna9_t <- as.data.frame(t(as.matrix(rna9)))
```

```{r}
# next two chunks for getting rid of 'unreplicated' cells 
cells <- as.data.frame(row.names(rna9_t))
names(cells) <- 'key'

cells %>% 
  separate(key, c('cell', 'rep'), remove = F) %>% 
  group_by(cell) %>% filter(n() == 2) %>% 
  ungroup() %>% select(key) -> cells_after_qc
```


```{r}
# filtering out rna9 after quality control
rna9_t %>% rownames_to_column() %>% 
  filter(rowname %in% cells_after_qc$key) %>% 
  column_to_rownames() -> rna9_f
```


```{r}
# Re arranging rna9_aqc moving cells as columns 
rna9_ft <- as.data.frame(t(as.matrix(rna9_f)))
rna9_ft %>% rownames_to_column("gene") -> rna9_ft 
```

```{r}
# Filtering out genes with less than 100 counts across cells 
rna9_ft %>%   
  arrange(gene) %>% rowwise() %>%  
  group_by(gene) %>% dplyr::summarise(m = sum(c_across(2:length(rna9_ft)-1))) %>%  
  filter(m > 100) %>% select(gene) -> exp_genes

rna_ftf <- as_tibble(rna9_ft[rna9_ft$gene %in% exp_genes$gene,])
dim(rna_ftf)
```

#Filtering out genes differentially expressed in specific stages
```{r}
rna_ftfc <- rna_ftf %>% filter(!gene %in% deg$gene)
```

Normalization
```{r}

# Step 1: divided each individual count by the total counts per cell, and multiply by 1000000 

rna_norm <- as_tibble(rna_ftfc$gene) # tibble ngenes x 1 for starting fillling with normalized data

for (i in 2:length(rna_ftfc)) {
  a <- (rna_ftfc[,i]/sum(rna_ftfc[,i]))*1000000
  rna_norm <- cbind(rna_norm, a)
  a <- NULL
}

rna_norm <- column_to_rownames(rna_norm, 'value') # for transposing

# Step 2: log transformation after add 11

rna_log <- log10(rna_norm[,1:length(rna_norm)] + 11 )

rna_log[1:5, 1:5]

rna_final <- as.data.frame(t(as.matrix(rna_norm)))
rna_final[1:5, 1:5]

```


```{r}
PCA = prcomp(rna_final, center=TRUE, scale.=FALSE)
summary(PCA)

PCA$x %>% 
  as.data.frame %>%
  ggplot(aes(x = PC1, y = PC2, size=100)) + 
  geom_point()+theme(legend.position = "None")

ggsave('pca2.jpeg')
```

#Figures assignment
Fig 1c
Fig2 a b c d
Fig3 a b d
Fig4 a b c d e
Fig5 c e

##Sean 3 a b d
```{r}

```


##Tanner 2 a c d




```{r}
#Dependencies
library("princurve")
library(data.table)
library(pheatmap)

```


```{r}

####Pseudotime

##Take Normalized genes and the filter with 2000 genes with highest variance and use for pseudotime. From fig. S3

 Anther_size = read.table("./Data/Updated_S1.txt", header= T)

#Transform data to merge files to use only 2000 genes
rna_final_Tanner= as.data.frame(t(rna_final)) 
s3= read.csv("./Data/aav6428_Nelms_Table-S3.csv", header=T)

#Add a column label to rownames
setDT(rna_final_Tanner, keep.rownames = "gene")[]

#Merge Files
Variance2000transcripts=inner_join(rna_final_Tanner,s3 )### MAke a combined table of all fig s3, order by variance and then slice 2000

y= Variance2000transcripts %>% arrange(desc("Gene score (% variance explained by pseudotime)")) %>% 
  select(1:258) %>% 
  top_n(2000)

 
#Remove gene id column
y$gene=NULL

#Trying the PCA with prcomp

PCA2= prcomp(y, center=TRUE, scale.=FALSE)
summary(PCA)
PCA2$x %>% 
  as.data.frame %>%
  ggplot(aes(x = PC1, y = PC2)) + 
  geom_point()

plot(PCA2)
names(PCA2)

#PCA 10 as a matrix while selecting the X values from the PCA
PCAX= as.matrix(PCA2$x [,1:10])
PCAX

I=principal_curve(PCAX)

#Principal Curve
plot(I$s, xlim=c(-20,10))

#Overlay principle curve for PCA in fig.1. Makes Fig. 2A
df1=as.data.frame(PCA$x)
df2=as.data.frame(I$s)

df1 %>% 
  ggplot(aes(x =PC1, y = PC2)) + 
  geom_point() +
  geom_smooth(data = df2, aes(x = PC1, y = PC2)) +
  theme(legend.position = "none")+xlim(-20,30)


ggsave('Pseudotime_PCA_overlay.jpeg')
```


```{r}
#### Heatmap  Fig. 2
#Sort by Anther size

ANTHER1=read.csv("./Data/with_anther_size", header=T)


ANTHER2=ANTHER1[order(ANTHER1$anther_size_mm),]
ANTHER2=as.tibble(ANTHER2)

#Remove cells that don't pass QC Result=144 cells
ANTHER3 <- filter(ANTHER2, cell %in% qc$`Cell Number`)
ANTHER4=t(ANTHER3)

#Remove rows X, cell, rep,anther_size_mm
ANTHER5=ANTHER4[-c(1,2,3,4),]
ANTHER6=as.data.frame(ANTHER5)

#Add row names as a column for sorting

setDT(ANTHER6, keep.rownames = "gene")[]

#Anther length with 3040 genes
Anther_length_w_genes=inner_join(ANTHER6,s3) # 3040 genes and 144 cells



#Remove gene names 
Anther_length_w_genes$gene=NULL

#Make heatmap ordered by anther length
Anther_length_w_genes1=(t(Anther_length_w_genes))

vst_cor<- as.matrix(cor(Anther_length_w_genes1, method="spearman"))

pheatmap(vst_cor,border_color="white",labels_col =NA,labels_row = "3040 genes", treeheight_row = 0, treeheight_col = 0)



```

##Yuru 4a,c,5c
```{r}

```

##Chiteri 1 c 2 b 5e
```{r}
# Quality control: PASS/FAIL
table1 <- readxl::read_excel('./Data/aav6428_Nelms_Table-S1.xlsx', sheet = 'Table S1')

#rename the columns..replace spaces with underscore and lower case
table1.1 <- table1 %>%  
  select_all(~gsub("\\s", "_", .)) %>% 
  select_all(tolower)

#rename anthersize column name
colnames(table1.1)[3] <- 'anther_size_mm'
colnames(table1.1)[1] <- 'cell'

#filter only cells that passed

table1.2 <- table1.1 %>%  filter(quality_control=="PASS") %>%  select(cell, anther_size_mm)

#Join two columns cell and rep and go wider

rna4_wide <- rna4 %>% pivot_wider(names_from = gene, values_from = count)

rna4_anther <- merge(rna4_wide, table1.2, by='cell') %>% select(cell, rep, anther_size_mm, everything()) %>% arrange(as.numeric(cell))

rna4_replicated <- rna4_anther %>% group_by(cell) %>% filter(n()==2) %>%  ungroup() 


#make anther_size_mm for grouping purposes
rna4_replicated$anther_size_mm <- as.character(rna4_replicated$anther_size_mm)

#filter whose sum across samples is > 100
rna4_filtered <- rna4_replicated[4:ncol(rna4_replicated)][, apply(rna4_replicated[4:ncol(rna4_replicated)],2,sum) > 100]


#we get 256 obs, 12912 genes, 10 shy of 12902 from the paper

#introduce the cell, rep, anther_size
rna4_x <- cbind(rna4_replicated[1:3], rna4_filtered )

#filter the 375 
# Deferentially expressed genes during specific stages
rna4_x2 <- rna4_x[, !names(rna4_x) %in% deg$gene] 
#we get 12544 against 12527 in the paper. We have 17 more??


#make anther_size_mm for grouping purposes
rna4_x2$anther_size_mm <- as.numeric(rna4_x2$anther_size_mm)

##normalize (uses columns)
#Uses row sums
rna4_x3 <- apply(rna4_x2[4:ncol(rna4_x2)], 1,function(i) log10(((i/sum(i))*1000000)+11 ))

rna4_log3 <- as.data.frame(t(rna4_x3))

#combine
rna4_log3 <- cbind(rna4_x2[,1:3], rna4_log3)

PCA_256 = prcomp(rna4_log3[, -c(1:3)], center=TRUE, scale.=FALSE)
summary(PCA_256)

#size=8 Will make the dots better. Place before colour

ggfortify::autoplot(PCA_256, data=rna4_x2, size=5, colour= 'anther_size_mm') + geom_line(aes(group = cell))+ scale_color_gradient2(low = "gray",high = "black", space = "Lab" )+theme_bw()

ggsave('pca2.png')




```
```{r}
##kmeans with cycle
#kmeans plot
set.seed(20)
#autoplot(kmeans(rna4_log3[, -c(1:3)], 6), data=rna4_x2 )

#Producing error currently
ggsave('cell_cycle_cluster.png')


```
```{r}
##kmeans with cycle
#kmeans plot
set.seed(20)
#autoplot(kmeans(rna4_log3[, -c(1:3)], 6), data=rna4_x2 )

#Producing error currently
ggsave('cell_cycle_cluster.png')

```



##Juan 4b d e

# Getting data for figures 4 b

```{r}
# Required data frames. From "Table.S4" we need gene names for Clusters 5 and 6  

Table.S4 <- readxl::read_excel('./Data/aav6428_Nelms_Table-S4.xlsx')
names(Table.S4) <- c('Gene', 'Category', 'Gene_name', 'Cluster', 'Meiocytes_Exp') 

cluster_5 <- Table.S4 %>% filter(Cluster == 5) %>% select(Gene)
cluster_6 <- Table.S4 %>% filter(Cluster == 6) %>% select(Gene)

# From Table.S1, we will use anther lenght

Table.S1 <- readxl::read_excel('./Data/aav6428_Nelms_Table-S1.xlsx', sheet = "Table S1")

Table.S1 %>% 
  filter(!`Cell cycle phase` == "N.D.") %>% droplevels() %>% 
  select(`Cell Number`, `Plant Number`, `Cell cycle phase`, `Anther size (mm)`) %>% 
  arrange(match(`Cell cycle phase`, c("G1", "G1/S", "S", "G2/M", "M", "mei")))


```
```{r}
rna_norm %>% rownames_to_column("gene") %>% 
  filter(gene %in% c(cluster_5$Gene, cluster_6$Gene)) %>% 
  column_to_rownames("gene") %>% t() %>% 
  as.data.frame() %>% 
  rownames_to_column('Cell') %>% 
  separate(Cell, into = c('Cell Number', 'Rep'), sep = "_") -> fig4 

fig4$`Cell Number` <- parse_number(fig4$`Cell Number`) 

fig4 %>%  
  left_join(., Table.S1[,c(1, 3)], by = "Cell Number") -> fig4t

library(patchwork)

fig4t$Stage <- cut(fig4t$`Anther size (mm)`, c(0, 0.75, 1.1, 1.3, 1.4, Inf), 
    labels = c('Mitotic expansion',
               "AR/PMC transition",
               'PMC/Pr1',
               "Pr1/PR2 transition",
               'End of Early meoitic prophase'))

fig4.b.1 <- fig4t %>% 
  ggplot(aes(x = `Anther size (mm)`, y = Zm00001d039133, color = Stage)) + 
  geom_point() +
  geom_smooth(color = 'red', se = F) +
  labs(title = "Afd1/Rec8") + xlab(NULL) + ylab("TPM") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5)) 

fig4.b.2 <- fig4t %>% 
  ggplot(aes(x = `Anther size (mm)`, y = Zm00001d025575, color = Stage)) + 
  geom_point() +
  geom_smooth(color = 'red', se = F) +
  labs(title = "Zip1") + xlab('Anther length (mm)') + ylab("TPM") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5))

fig4b <- (fig4.b.1 / fig4.b.2) + plot_layout(guides = 'collect')

ggsave('./Figures/Figure4_b.png', plot = fig4b)

```
Additional figures for 2D
```{r}
############################   Extra figures 2.D  ##################################

fig2 <- as.tibble(rna9_ft)

fig2df <- as.tibble(fig2$gene)
for (i in 2:length(fig2)) {
  a <- (fig2[,i]/sum(fig2[,i]))*1000000
  fig2df <- cbind(fig2df, a)
  a <- NULL
}

fig2df <- fig2df %>%  
  filter(value %in% c('Zm00001d027037', 'Zm00001d039101', 'Zm00001d009343', 'Zm00001d050069', 'Zm00001d013377')) 

fig2df %>% 
  column_to_rownames("value") %>% t() %>% 
  as.data.frame() %>% 
  rownames_to_column('Cell') %>% 
  separate(Cell, into = c('Cell Number', 'Rep'), sep = "_") -> fig2ddf

fig2ddf$`Cell Number` <- parse_number(fig2ddf$`Cell Number`) 

fig2ddf %>%  
  left_join(., Table.S1[,c(1, 3)], by = "Cell Number") -> fig2ddf_sum

fig2ddf_sum$Stage <- cut(fig2ddf_sum$`Anther size (mm)`, c(0, 0.75, 1.1, 1.3, 1.4, Inf), 
    labels = c('Mitotic expansion',
               "AR/PMC transition",
               'PMC/Pr1',
               "Pr1/PR2 transition",
               'End of Early meoitic prophase'))

Unk1 <- 
  fig2ddf_sum %>%
  ggplot(aes(x = `Anther size (mm)`, y = Zm00001d027037, color = Stage)) + 
  geom_point() +
  geom_smooth(color = 'red', se = F) +
  labs(title = "Zip1") + xlab(NULL) + ylab("TPM") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0))

  
Rmf <- fig2ddf_sum %>% 
  ggplot(aes(x = `Anther size (mm)`, y = Zm00001d039101, color = Stage)) + 
  geom_point() +
  geom_smooth(color = 'red', se = F) +
  labs(title = "Rmf") + xlab(NULL) + ylab("TPM") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0))

Rps4a <- fig2ddf_sum %>% 
  ggplot(aes(x = `Anther size (mm)`, y = Zm00001d009343, color = Stage)) + 
  geom_point() +
  geom_smooth(color = 'red', se = F) +
  labs(title = "Rps4a") + xlab(NULL) + ylab("TPM") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0))

Trps8 <- 
  fig2ddf_sum %>%
  ggplot(aes(x = `Anther size (mm)`, y = Zm00001d050069, color = Stage)) + 
  geom_point() +
  geom_smooth(color = 'red', se = F) +
  labs(title = "Trps8") + xlab(NULL) + ylab("TPM") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0))

Unk2 <- 
  fig2ddf_sum %>%
  ggplot(aes(x = `Anther size (mm)`, y = Zm00001d013377, color = Stage)) + 
  geom_point() +
  geom_smooth(color = 'red', se = F) +
  labs(title = "Unk2") + xlab('Anther length (mm)') + ylab("TPM") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0))

fig2dextra <- (Unk1 / Rmf / Trps8 / Unk2) + plot_layout(guides = 'collect')

ggsave('./Figures/Figure2_d.png', plot = fig2dextra)
```

