---
title: "BCB_Project"
author: "Chiteri and Tanner"
date: "4/4/2021"
output: html_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
#Dependencies
library("tidyverse")
library("factoextra")
library("FactoMineR")
library("readxl")

```

```{r}
#Set working directory
setwd('.')
getwd()
list.files()
```

Reading data

```{r}
# Counts
rna <- read_csv('./GSE121039_AR_scRNAseq_transcript_counts.csv')
rna <- rename(rna, gene = X1) #rename first column

# Quality control: PASS/FAIL
qc <- readxl::read_excel('./aav6428_Nelms_Table-S1.xlsx', sheet = "Table S1")
names(qc)[12] <- 'QC' 
qc <- qc[qc$QC == "PASS", "Cell Number"]

# Deferentially expressed genes during specific stages
deg <- readxl::read_excel('./aav6428_Nelms_Table-S2.xlsx')[,1:2]
names(deg) <- c('gene', 'phase')
deg <- deg %>% filter(phase != "G1") %>% select(gene)

head(rna)
head(qc)
```

```{r}
#get into long format
rna1 <- pivot_longer(rna, cols = starts_with('Cell'))
```


```{r}
#separate 'name' into 4 columns, 
#rename the columns
#select the columns

rna3 <- separate(rna1, name, into = paste0("y",1:4), sep = " ") %>% 
  select(!c(y1, y3)) %>% rename(c('cell' = y2, 'rep'= y4, 'count' = value))

```


```{r}
#Filter cells based on QC
rna4 <- filter(rna3, cell %in% qc$`Cell Number`)
dim(rna4)
head(rna4)

paste0("Total number of cells after Quality Control: ", length(unique(as.numeric(rna4$cell))))
```


``` {r}
##convert cell & rep into numeric & Count unique cells.
n.cells2 <- unique(as.numeric(rna4$cell)) #now we have 144 cells

#rna4 contains all cells that passed QC and were used in the analysis
```


``` {r}

##dataset to use for e.g. Fig. 2A, Fig S4). We need 128 cells
#pooling cells using sum
rna5 <- rna4 %>%  group_by(gene,cell,) %>% summarise(n=n(), sum = sum(count))
paste("There are ", length(unique(as.numeric(rna5$cell))), " unique cells in rna5")
```

``` {r}
rna6 <- filter(rna5, n > 1)

paste0("There are ", length(unique(as.numeric(rna6$cell))), " cells with a technical replicate in rna6.")
```


``` {r}
#Join two columns cell and rep and go wider

rna7 <- rna4 %>%  
  unite("cell",c(cell, rep), sep = "_") %>% 
  pivot_wider(names_from = cell, values_from = count, names_prefix = 'cell')

dim(rna7)
```


``` {r}
#remove gene title
rna9 <- rna7 %>% remove_rownames %>% column_to_rownames(var = "gene")

dim(rna9)

rna7[1:5, 1:5]
rna9[1:5, 1:4]

```

```{r}
# transpose data frame so cells are row and genes are columns
rna9_t <- as.data.frame(t(as.matrix(rna9)))
```

```{r}
# next two chunks for getting rid of 'unreplicated' cells 
cells <- as.data.frame(row.names(rna9_t))
names(cells) <- 'key'

cells %>% 
  separate(key, c('cell', 'rep'), remove = F) %>% 
  group_by(cell) %>% filter(n() == 2) %>% 
  ungroup() %>% select(key) -> cells_after_qc
```


```{r}
# filtering out rna9 after quality control
rna9_t %>% rownames_to_column() %>% 
  filter(rowname %in% cells_after_qc$key) %>% 
  column_to_rownames() -> rna9_f
```


```{r}
# Re arranging rna9_aqc moving cells as columns 
rna9_ft <- as.data.frame(t(as.matrix(rna9_f)))
rna9_ft %>% rownames_to_column("gene") -> rna9_ft 
```

```{r}
# Filtering out genes with less than 100 counts across cells 
rna9_ft %>%   
  arrange(gene) %>% rowwise() %>%  
  group_by(gene) %>% dplyr::summarise(m = sum(c_across(2:length(rna9_ft)-1))) %>%  
  filter(m > 100) %>% select(gene) -> exp_genes

rna_ftf <- as.tibble(rna9_ft[rna9_ft$gene %in% exp_genes$gene,])
dim(rna_ftf)
```
Filtering out genes differentially expressed in specific stages
```{r}
rna_ftfc <- rna_ftf %>% filter(!gene %in% deg$gene)
```

Normalization
```{r}

# Step 1: divided each individual count by the total counts per cell, and multiply by 1000000 

rna_norm <- as.tibble(rna_ftfc$gene) # tibble ngenes x 1 for starting fillling with normalized data

for (i in 2:length(rna_ftfc)) {
  a <- (rna_ftfc[,i]/sum(rna_ftfc[,i]))*1000000
  rna_norm <- cbind(rna_norm, a)
  a <- NULL
}

rna_norm <- column_to_rownames(rna_norm, 'value') # for transposing

# Step 2: log transformation after add 11

rna_norm <- log10(rna_norm[,1:length(rna_norm)] + 11 )

rna_norm[1:5, 1:5]

rna_final <- as.data.frame(t(as.matrix(rna_norm)))
rna_final[1:5, 1:5]

```


```{r}
PCA = prcomp(rna_final, center=TRUE, scale.=TRUE)
summary(PCA)

PCA$x %>% 
  as.data.frame %>%
  ggplot(aes(x = PC1, y = PC2)) + 
  geom_point()

```

#Figures assignment
Fig 1c
Fig2 a b c d
Fig3 a b d
Fig4 a b c d e
Fig5 c e

##Sean 3 a b d
```{r}

```


##Tanner 2 a c d
```{r include=FALSE}
library("princurve")
install.packages('PRcurve')
suppressWarnings(suppressPackageStartupMessages(library(ggthemes)))
suppressWarnings(suppressPackageStartupMessages(library(ggplot2)))
suppressWarnings(suppressPackageStartupMessages(library(scran)))
suppressWarnings(suppressPackageStartupMessages(library(cowplot)))

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
library(slingshot)

library(monocle)
library("princurve")
library(dplyr)

##PCA for pairwise comparison of replicates
 
library(readr)
library(tidyverse)

Anther_size = read.table("Updated_S1.txt", header= T)



Anther_size1 = read.csv("Updated_S1.csv", header= T)

rna4a= as.integer(rna4$cell)

Anther_sort=right_join(rna4, Anther_size, "cell")




#Identify 2000 genes with highest variance and use for pseudotime. From fig. S3
variance2000 = read.table("2000genE_s3.txt", header= T)

Variance2000transcripts=right_join(rna7,variance2000, "gene")

#Trying the PCA with prcomp
#remove gene title
rna10 <- Variance2000transcripts %>% remove_rownames %>% column_to_rownames(var = "gene")
PCA= prcomp(rna10[, c(2:272)], center=TRUE, scale.=TRUE)
summary(PCA)
plot(PCA)

##Allows us to see the variable in the PCA result
# "X" is the principal components of interest *Source:https://cmdlinetips.com/2019/04/introduction-to-pca-with-r-using-prcomp/
names(PCA)

PCA[1:10]

#PCA 10 as a matrix while selecting the X values from the PCA
PCAX= as.matrix(PCA$x, (c[,1:10]))
PCAX

#Correct way to only use PC10? I don't think so but I am unsure

I=principal_curve(PCAX, smoother="smooth.spline")
print(I)

J=PCA$x %>% 
  as.data.frame %>%
  ggplot(aes(x=PC1,y=PC2)) + geom_point(size=4) +
  theme_bw(base_size=32)
J

autoplot(PCA)

plot(I)
plot(I$lambda)
plot(PCA$x) # any of the PCA names can be called here (where x is)








####
library(stringr)
library(dplyr)

ls()
rm(list=setdiff(ls(), "rna4"))

check_reps <- rna4 %>% group_by(gene, cell) %>% summarise(n = n()) 
head(check_reps)
unique(check_reps$n)
t
str(rna4)

check_reps <- check_reps %>% filter(n == 1)


to_remove <- paste0(check_reps$gene, check_reps$cell)
to_remove[1:5]

rna4$index <- paste0(rna4$gene, rna4$cell)

pca_data <- rna4[-match(to_remove,rna4$index),]

dim(pca_data)
dim(rna4)
head(pca_data)



check_reps <- pca_data %>% group_by(gene, cell) %>% summarise(n = n()) 
check_reps[which(check_reps$n != 2),]


pca_data_final <- data.frame(R1 = pca_data$count[which(pca_data$rep == 1)],
                             R2 = pca_data$count[which(pca_data$rep == 2)])


sum(pca_data$rep == 1)
sum(pca_data$rep == 2)

#Using Monocle


##Yuru 4a,c,5c
```{r}

```

##Chiteri 1 c 2 b 5e
```{r}

```

##Juan 4b d e
```{r}

```


